src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp

```diff
@@ -0,0 +1,233 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Domain/FunctionsOfTime/QuaternionFunctionOfTime.hpp"
+
+// BoostMultiArray is used internally in odeint, so BoostMultiArray MUST be
+// included before odeint
+#include "DataStructures/BoostMultiArray.hpp"
+
+#include <boost/numeric/odeint.hpp>
+#include <pup_stl.h>
+
+#include "Domain/FunctionsOfTime/QuaternionHelpers.hpp"
+#include "NumericalAlgorithms/OdeIntegration/OdeIntegration.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/ErrorHandling/Error.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/MakeArray.hpp"
+#include "Utilities/StdHelpers.hpp"
+
+namespace domain::FunctionsOfTime {
+template <size_t MaxDeriv>
+QuaternionFunctionOfTime<MaxDeriv>::QuaternionFunctionOfTime(
+    double t, std::array<DataVector, 1> initial_quat_func,
+    std::array<DataVector, MaxDeriv + 1> initial_omega_func,
+    double expiration_time) noexcept
+    : stored_quaternions_and_times_{{t, std::move(initial_quat_func)}},
+      omega_f_of_t_(t, initial_omega_func, expiration_time),
+      expiration_time_(expiration_time) {}
+
+template <size_t MaxDeriv>
+std::unique_ptr<FunctionOfTime> QuaternionFunctionOfTime<MaxDeriv>::get_clone()
+    const noexcept {
+  return std::make_unique<QuaternionFunctionOfTime>(*this);
+}
+
+template <size_t MaxDeriv>
+void QuaternionFunctionOfTime<MaxDeriv>::pup(PUP::er& p) {
+  FunctionOfTime::pup(p);
+  p | stored_quaternions_and_times_;
+  p | expiration_time_;
+  p | omega_f_of_t_;
+}
+
+template <size_t MaxDeriv>
+void QuaternionFunctionOfTime<MaxDeriv>::update(
+    const double time_of_update, const DataVector& updated_max_deriv,
+    const double next_expiration_time) noexcept {
+  omega_f_of_t_.update(time_of_update, updated_max_deriv, next_expiration_time);
```

Comment:
I had a const reference because `omega_f_of_t_.update` moves `updated_max_deriv` so I didn't want to make an extra copy.

Reply:
That's why you take by value and move things, actually.  There are two cases: the caller is giving up ownership of the passed in resources (either because the passed thing is a temporary or because it was explicitly moved), or the owner is not.  (These cases have names, but I can never keep them straight.)

| Ownership transfer | update() argument type | origin of update() argument | origin of `omega_f_of_t_.update()` argument |
| ------------------ | ---------------------- | --------------------------- | ------------------------------------------- |
| no                 | `const DataVector&`    | reference                   | copy                                        |
| no                 | `DataVector`           | copy                        | move                                        |
| yes                | `const DataVector&`    | reference                   | copy                                        |
| yes                | `DataVector`           | move                        | move                                        |

So if the caller is not giving up ownership, there is one copy either way.  (The by-value version does incur an extra move, but those are generally assumed to be cheap.)  But, if the caller *is* giving up ownership, taking by value avoids a copy.  (If the argument of `omega_f_of_t_.update()` were a reference, on the other hand, you would get the opposite conclusion and want to take by reference here as well, but that result is more intuitive.)